src/tests/ftest/util/dfuse_utils.py

```diff
@@ -173,21 +197,35 @@ def stop(self):
         """Stop dfuse
         Raises:
             CommandFailure: In case dfuse stop fails
+
+        Try to stop dfuse.  Try once nicely by using fusermount, then if that
+        fails try to pkill it to see if that works.  Abort based on the result
+        of the fusermount, as if pkill is necessary then dfuse itself has
+        not worked correctly.
+
+        Finally, try and remove the mount point, and that itself should work.
         """
+        self.log.info('Stopping dfuse at %s', self.mount_dir.value)
 
+        if self.mount_dir.value is None:
+            return
         cmd = "if [ -x '$(command -v fusermount)' ]; "
         cmd += "then fusermount -u {0}; else fusermount3 -u {0}; fi".\
                format(self.mount_dir.value)
         ret_code = general_utils.pcmd(self.hosts, cmd, timeout=30)
-        self.remove_mount_point()
-        if 0 not in ret_code:
+        if len(ret_code) > 1 or 0 not in ret_code:
             error_hosts = NodeSet(
                 ",".join(
                     [str(node_set) for code, node_set in ret_code.items()
                      if code != 0]))
+
+            cmd = "pkill dfuse --signal KILL"
```

Comment:
Just confirmed that this works.  Had two of these hung up at the same time even and it worked on both.

The problem is though, that this need to `pkill` does not result in any kind of test failure.  It should, shouldn't it?  Going to -1 this until we resolve that question and update the PR if it should cause a test failure.

Reply:
Yes, this should result in a test failure.

When I've used python unittesting in the past (which I think Avacaodo is still just a fancy wrapper around) of the test fails you get a fail, and if the cleanup fails you get an error, so the exception raised here should cause it to error.

It's tricky though, I think the correct thing (and this applies to daos servers/agents as well as dfuse) would be to stop it as part of the test, and do any log-analysis steps there as well so that any errors reported could be correctly fed back into CI as test failures - but this would probably be quite an intrusive change.

Did the pkill clear the mount, or did it leave a stale mount that would cause problems with future dfuse tests, did you happen to check /proc/mounts after killing it?

Reply:
Hrm.  So on the blocked run I tested this `pkill` on, the run actually ended up a complete pass, so the test following was successful, indicating that the `pkill` is effective in cleaning up the node.  I did not pay attention to the mount point after-ward though.

So, it's still concerning -- enough to not land this until we get to the bottom of it, IMHO -- to ensure that the need to `pkill` is resulting in a failed test run.  That's not what my test showed.

Avocado is actually a full re-implementation of a subset of the python unittest framework rather than being a wrapper around actual python-unittest.  It is supposed to implement the behaviour that raised exceptions mean test failure, so it's interesting that this didn't happen in my test above.

We might need to wait for another blocked run and experiment more before landing this patch to avoid (unintentionally of course) papering over the issue.

Reply:
If the test blocked, needed to be cleaned up with pkill but then the PR passed all testing then yes, that is a worry.  Do you have a link that I can login to?

What does happen if there's an exception raised in (what I assume is) the cleanup handler, I know it's logged as being ignored but does it at least result in a test failure?  i think I wrote this code to raise an exception on pkill so we might be OK with this new code but I'll need to check the logs.

Reply:
I don't think we are OK.  Your added exception raising is just raising exceptions in the same was as the existing code which is ignoring them:

```
2020-03-09 01:09:09,316 output           L0655 DEBUG| Exception command_utils.CommandFailure: CommandFailure('Error removing the /tmp/daos_dfuse dfuse mount point on the following hosts: wolf-[54,146-147]',) in <bound method Dfuse.__del__ of <dfuse_utils.Dfuse object at 0x7fe96465d710>> ignored
```

But I think this issue is bigger than this PR and so have opened [DAOS-4299](https://jira.hpdd.intel.com/browse/DAOS-4299) about it.

The unfortunate thing about the instance of this problem in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/1820/artifact/Functional/job-2020-03-09T01.02-3d2599d-io-mdtest_small/job.log is that the mdtest in that run failed for other reasons:

```
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] error could not create directory '/tmp/daos_dfuse/test-dir.0-0/mdtest_tree.1.0/'
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] error could not create directory '/tmp/daos_dfuse/test-dir.0-0/mdtest_tree.3.0/'
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] error could not create directory '/tmp/daos_dfuse/test-dir.0-0/mdtest_tree.5.0/'
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] error could not create directory '/tmp/daos_dfuse/test-dir.0-0/mdtest_tree.7.0/'
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] 03/09/2020 01:09:10: Process 1: FAILED in create_remove_directory_tree, Unable to create directory /tmp/daos_dfuse/test-dir.0-0/mdtest_tree.1.0/mdtest_tree.1.1/: No such file or directory
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] 03/09/2020 01:09:10: Process 3: FAILED in create_remove_directory_tree, Unable to create directory /tmp/daos_dfuse/test-dir.0-0/mdtest_tree.3.0/mdtest_tree.3.1/: No such file or directory
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] 03/09/2020 01:09:10: Process 5: FAILED in create_remove_directory_tree, Unable to create directory /tmp/daos_dfuse/test-dir.0-0/mdtest_tree.5.0/mdtest_tree.5.1/: No such file or directory
2020-03-09 01:09:04,977 process          L0479 DEBUG| [stdout] 03/09/2020 01:09:10: Process 7: FAILED in create_remove_directory_tree, Unable to create directory /tmp/daos_dfuse/test-dir.0-0/mdtest_tree.7.0/mdtest_tree.7.1/: No such file or directory
2020-03-09 01:09:05,169 process          L0499 INFO | Command 'export DAOS_SINGLETON_CLI=1; export DAOS_SVCL=1; export DAOS_POOL=90DC5921-89EB-4746-A10B-F283FE06B096; export D_LOG_FILE=/var/tmp/daos_testing/mdtest_small.py-1_daos_client.log; export CRT_ATTACH_INFO_PATH=/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_master/install/tmp; export MPI_LIB=""; export FI_PSM2_DISCONNECT=1; /usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_iKhSQ_/1-._io_mdtest_small.py_MdtestSmall.test_mdtest_small_dfuse-hosts-np_8-pool-server_config-5372/hostfile27944 -np 8 mdtest -a POSIX -b 2 -z 5 -u -I 10 -i 1 -e 0 -d /tmp/daos_dfuse -w 0' finished with 1 after 1.18693518639s
2020-03-09 01:09:05,169 command_utils    L0318 ERROR| Error occurred running 'export DAOS_SINGLETON_CLI=1; export DAOS_SVCL=1; export DAOS_POOL=90DC5921-89EB-4746-A10B-F283FE06B096; export D_LOG_FILE=/var/tmp/daos_testing/mdtest_small.py-1_daos_client.log; export CRT_ATTACH_INFO_PATH=/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_master/install/tmp; export MPI_LIB=""; export FI_PSM2_DISCONNECT=1; /usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_iKhSQ_/1-._io_mdtest_small.py_MdtestSmall.test_mdtest_small_dfuse-hosts-np_8-pool-server_config-5372/hostfile27944 -np 8 mdtest -a POSIX -b 2 -z 5 -u -I 10 -i 1 -e 0 -d /tmp/daos_dfuse -w 0': Command 'export DAOS_SINGLETON_CLI=1; export DAOS_SVCL=1; export DAOS_POOL=90DC5921-89EB-4746-A10B-F283FE06B096; export D_LOG_FILE=/var/tmp/daos_testing/mdtest_small.py-1_daos_client.log; export CRT_ATTACH_INFO_PATH=/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_master/install/tmp; export MPI_LIB=""; export FI_PSM2_DISCONNECT=1; /usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_iKhSQ_/1-._io_mdtest_small.py_MdtestSmall.test_mdtest_small_dfuse-hosts-np_8-pool-server_config-5372/hostfile27944 -np 8 mdtest -a POSIX -b 2 -z 5 -u -I 10 -i 1 -e 0 -d /tmp/daos_dfuse -w 0' failed (rc=1)
2020-03-09 01:09:05,169 mdtest_test_base L0190 ERROR| Mdtest Failed: Error occurred running 'export DAOS_SINGLETON_CLI=1; export DAOS_SVCL=1; export DAOS_POOL=90DC5921-89EB-4746-A10B-F283FE06B096; export D_LOG_FILE=/var/tmp/daos_testing/mdtest_small.py-1_daos_client.log; export CRT_ATTACH_INFO_PATH=/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_master/install/tmp; export MPI_LIB=""; export FI_PSM2_DISCONNECT=1; /usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_iKhSQ_/1-._io_mdtest_small.py_MdtestSmall.test_mdtest_small_dfuse-hosts-np_8-pool-server_config-5372/hostfile27944 -np 8 mdtest -a POSIX -b 2 -z 5 -u -I 10 -i 1 -e 0 -d /tmp/daos_dfuse -w 0': Command 'export DAOS_SINGLETON_CLI=1; export DAOS_SVCL=1; export DAOS_POOL=90DC5921-89EB-4746-A10B-F283FE06B096; export D_LOG_FILE=/var/tmp/daos_testing/mdtest_small.py-1_daos_client.log; export CRT_ATTACH_INFO_PATH=/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_master/install/tmp; export MPI_LIB=""; export FI_PSM2_DISCONNECT=1; /usr/lib64/mpich/bin/mpirun -hostfile /var/tmp/avocado_iKhSQ_/1-._io_mdtest_small.py_MdtestSmall.test_mdtest_small_dfuse-hosts-np_8-pool-server_config-5372/hostfile27944 -np 8 mdtest -a POSIX -b 2 -z 5 -u -I 10 -i 1 -e 0 -d /tmp/daos_dfuse -w 0' failed (rc=1)
```

so it's difficult to determine from this run if the raised exception that is being ignored (i.e above) would have resulted in a passing test failing due to the teardown failure.  I am pretty certain the run I `pkill`ed a `dfuse` in on Friday did end up with a full pass, so the raised exception did not fail it.
```